barretenberg/cpp/src/barretenberg/commitment_schemes/gemini/gemini_impl.hpp

```diff
@@ -49,19 +49,37 @@ std::vector<typename GeminiProver_<Curve>::Claim> GeminiProver_<Curve>::prove(
     const std::shared_ptr<CommitmentKey<Curve>>& commitment_key,
     const std::shared_ptr<Transcript>& transcript,
     RefSpan<Polynomial> concatenated_polynomials,
-    const std::vector<RefVector<Polynomial>>& groups_to_be_concatenated)
+    const std::vector<RefVector<Polynomial>>& groups_to_be_concatenated,
+    bool HasZK)
 
 {
     size_t log_n = numeric::get_msb(static_cast<uint32_t>(circuit_size));
     size_t n = 1 << log_n;
 
-    Fr rho = transcript->template get_challenge<Fr>("rho");
-
     // Compute batched polynomials
     Polynomial batched_unshifted(n);
-    Polynomial batched_to_be_shifted = Polynomial::shiftable(1 << log_n);
+    Polynomial batched_to_be_shifted = Polynomial::shiftable(n);
+
+    // To achieve ZK, we mask the batched polynomial by a random polynomial of the same size
+    if (HasZK) {
+        batched_unshifted = Polynomial::random(n);
```

Comment:
Before: initialize poly to zero then start to accumulate f_polynomials[0]:
More efficient: initialize with `f_polynomials[0]` and start the loop at index 1.

After: initialize poly to zero then start to accumulate f_polynomials[0] OR overwrite with random.
More efficient: conditionally initialize as in 

```
Polynomial batched_unshifted =  has_zk ?  Polynomial::random(n) : f_polynomials[0];
size_t starting_idx = has_zk ? 0 : 1;
...
for (size_t i = starting_idx; i < f_polynomials.size(); i++) {
...
```
But then you have to similarly define a conditional starting point for the loop. Anyway, you're not making things worse so don't worry about it, just remarking.

Reply:
thanks for the suggestion! it breaks some edge case tests in Gemini though, namely when we open shifts without unshifted (quite unrealistic). adding another check (f_polynomials.size()>0) would be somewhat ugly